lib/ecto/type.ex

```diff
@@ -534,6 +561,8 @@ defmodule Ecto.Type do
   defp of_base_type?(:string, term), do: is_binary(term)
   defp of_base_type?(:map, term),    do: is_map(term) and not Map.has_key?(term, :__struct__)
 
+  defp of_base_type?({:map, model}, %{__struct___: struct}), do: model == struct
```

Comment:
The second argument can be a string... and it will explode in this case. Let's add tests for `{:map, Model}` to `Ecto.TypeTest` please. the proper solution for this case would be something like:

    defp of_base_type?({:map, struct}, %{__struct___: struct}), do: true
    defp of_base_type?({:map, _struct}, _), do: false